buildSrc/src/main/minimumRuntime/org/elasticsearch/gradle/LoggedExec.java

```diff
@@ -19,21 +26,52 @@
 @SuppressWarnings("unchecked")
 public class LoggedExec extends Exec {
 
+    // helper to allow lambda creation using Files methods
+    private interface ThrowingStringSupplier {
+        String get() throws Exception;
+    }
+
+    private Property<Boolean> useFileBuffer;
+    
     public LoggedExec() {
-        ByteArrayOutputStream output = new ByteArrayOutputStream();
-        ByteArrayOutputStream error = new ByteArrayOutputStream();
+        this.useFileBuffer = getProject().getObjects().property(Boolean.class);
+        this.useFileBuffer.set(false); // default to in memory
         if (getLogger().isInfoEnabled() == false) {
-            setStandardOutput(output);
-            setErrorOutput(error);
-            setIgnoreExitValue(true);
-            doLast((unused) -> {
+            getProject().afterEvaluate(p -> {
+                OutputStream stdout;
+                OutputStream stderr;
+                ThrowingStringSupplier readStdout;
+                ThrowingStringSupplier readStderr;
+
+                try {
+                    if (useFileBuffer.get()) {
+                        String outputdir = p.getBuildDir() + "/buffered-output/" + this.getName();
+                        p.mkdir(outputdir);
+                        stdout = new FileOutputStream(outputdir + "/stdout");
+                        stderr = new FileOutputStream(outputdir + "/stderr");
+                        readStdout = () -> Files.readString(Paths.get(outputdir + "/stdout"));
+                        readStderr = () -> Files.readString(Paths.get(outputdir + "/stderr"));
+                    } else {
+                        stdout = new ByteArrayOutputStream();
+                        stderr = new ByteArrayOutputStream();
+                        readStdout = () -> ((ByteArrayOutputStream) stdout).toString(StandardCharsets.UTF_8);
+                        readStderr = () -> ((ByteArrayOutputStream) stderr).toString(StandardCharsets.UTF_8);
+                    }
+                } catch (IOException e) {
+                    throw new RuntimeException(e);
+                }
+
+                setStandardOutput(stdout);
```

Comment:
FWIW, I find separating stdout and stderr generally unhelpful in practice. It makes it really hard to correlate messages when the executable writes to both streams. My preference is simply to write both to the same stream and output the same way. If we _really_ care about which stream the output came from we can prefix the log lines like we do for test logging.